## Overview of the folder `06-figures_tables`

This folder contains the final figures and tables presented in this study.

### Figures:

- `Figure1.pdf`: main figure containing an overview map of all samples, the DNA molecule length and
  contig length distributions for the samples EMN001, PES001, and PLV001, the overview of the
  reconstructed MAGs for the five samples with a HQ Chlorobiaceae MAG, and the aDNA damage for two
  MAGs of EMN001

### Supplementary Tables:

- `Dataset_S1.xlsx`: XLSX file containing the tables holding the sample and sequencing data
  information as well as the number of DNA moleculues that were available for *de novo* assembly per
  sample
- `Dataset_S2.xlsx`: XLSX file containing the results of the evaluation of the *de novo* assembly
  performance
- `Dataset_S3.xlsx`: XLSX file containing the number of MAGs after binning and the results of the
  the quality evaluation before and after the automatic refinement of the metaWRAP refined MAGs
- `Dataset_S4.xlsx`: XLSX file containing the results of the alignment of the short-read sequencing
  data against the assembled contigs of EMN001 for all dental calculus samples and the Neanderthal
  dental calculus samples previously published by Weyrich *et al.* (2017), and the results on the
  genotype calling along the EMN001 Chlorobiaceae MAG using snpAD
- `Dataset_S5.txt`: XLSX file containing all the assembly accessions of the Chlorobiales genomes
  as downloaded from the NCBI assembly database on the 2nd of November 2021 and the ones used
  for the dreplicated trees and for Figure2.
- `Dataset_S6.xlsx` : XLXS file containing all the results of the functional annotation of the ancient
  bins using EggNog and diamond alignment.
- `Dataset_S7.xlsx`: XLXS file containing the information on the sequencing data of the control
  samples from the site El Miron and the laboratory negative controls with the the table containing
  the results on the alignment of these data against the contigs assembled from EMN001 
- `Dataset_S9.xlsx` : XLXS file containing the network analysis of the butyrolactone BGCs as generated
  by bigscape and a summary of the GCFs and GCCs estimated.


### Supplementary Figures:

- `FigureS_ASMB01.pdf`: Supplementary figure summarising the results of the substitution types and
  minor allele frequencies that were observed when correcting the contig sequences using freeBayes
- `FigureS_ASMB02.pdf`: Supplementary figure summarising the results of the evaluation of the *de
  novo* assembly performance with respect to the contig lengths
- `FigureS_ASMB03.pdf`: Supplementary figure summarising the results of pyDamage on contig level
- `FigureS_ASMB04.pdf`: Supplementary figure summarising the quality estimates of the MAGs refined
  by metaWRAP estimated using checkM, BUSCO, and GUNC
- `FigureS_ASMB05.pdf`: Supplementary figure summarising the quality estimates of the MAGs refined
  by metaWRAP estimated using checkM and GUNC after automatically refining them using MMSeqs2
- `FigureS_ASMB06.pdf`: Supplementary figure summarising the taxonomic assignmnet of the MAGs
  against the GTDB
- `FigureS_ASM007.pdf`: Supplementary figure summarising of the ancient DNA damage of the samples
  with a HQ Chlorobium MAG against the representative MAG of the genus Chlorobium and Flexilinea of
  EMN001
