## Overview of the folder `06-figures_tables`

This folder contains the final figures and tables presented in this study.

### Figures:

- `Figure1.pdf`: main figure containing an overview map of all samples, the DNA molecule length and
  contig length distributions for the samples EMN001, PES001, and PLV001, the overview of the
  reconstructed MAGs for the five samples with a HQ Chlorobiaceae MAG, and the aDNA damage for two
  MAGs of EMN001

### Supplementary Tables:

- `Dataset_S1.xlsx`: XLSX file containing the tables holding the sample and sequencing data
  information as well as the number of DNA moleculues that were available for *de novo* assembly per
  sample
- `Dataset_S2.xlsx`: XLSX file containing the results of the evaluation of the *de novo* assembly
  performance
- `Dataset_S4.xlsx`: XLSX file containing the results of the alignment of the short-read sequencing
  data against the assembled contigs of EMN001 for all dental calculus samples and the Neanderthal
  dental calculus samples previously published by Weyrich *et al.* (2017), and the results on the
  genotype calling along the EMN001 Chlorobiaceae MAG using snpAD

### Supplementary Figures:

- `FigureS_ASMB01.pdf`: Supplementary figure summarising the results of the substitution types and
  minor allele frequencies that were observed when correcting the contig sequences using freeBayes
- `FigureS_ASMB02.pdf`: Supplementary figure summarising the results of the evaluation of the *de
  novo* assembly performance with respect to the contig lengths
- `FigureS_ASMB03.pdf`: Supplementary figure summarising the results of pyDamage on contig level
- `FigureS_ASMB04.pdf`: Supplementary figure summarising the quality estimates of the MAGs refined
  by metaWRAP estimated using checkM, BUSCO, and GUNC
