{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Project**: Natural products from the Palaeolithic\n",
    "\n",
    "## **Section**: Genome assemblies\n",
    "\n",
    "Anan Ibrahim, 01.01.2022\n",
    "\n",
    "**Contents**\n",
    " - **Step1**: Create conda envirorment with required dependencies if not already installed \n",
    " - **Step2**: Download sequencing results from eurofins using commandline\n",
    " - **Step3**: Genome assembly"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step1**: Create conda envirorment with required dependencies if not already installed \n",
    "\n",
    "##########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# All conda envs can be found in EMN001_Paleofuran/02-scripts/ENVS_*.yml\n",
    "conda env create -f plasmid_assembly.yml\n",
    "conda env create -f single_genome_assembly.yml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step2**: Download sequencing results from eurofins using commandline\n",
    "\n",
    "##########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir /Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly && cd \"$_\"\n",
    "mkdir /Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly/Input && cd \"$_\"\n",
    "mkdir /Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly/Output && cd \"$_\"\n",
    "mkdir /Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly/Scripts && cd \"$_\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**PROJECT SPECIFIC**\n",
    "\n",
    "*Manually:* Add the fastq sequences of the genomesretrieved from eurofins (in the downloads folder below) in folders named by samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir /Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly/Input/p_protegens && cd \"$_\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir /Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly/Input/p_temperata && cd \"$_\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir /Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly/eurofins-downloads && cd \"$_\"\n",
    "\n",
    "wget -m --ftp-user=########### --ftp-password=######### ftp://ftp.gatc-biotech.com/2021-12-21/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Manually:* Add the ref sequences of the genomes retrieved from eurofins in folders named by samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir /Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly/Input/p_protegens_ref_seq && cd \"$_\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir /Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly/Input/p_temperata_ref_seq && cd \"$_\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How to deal with jupyter notebook on your local computer\n",
    "\n",
    " - If miniconda is not installed, install miniconda in your local directory. Close and reopen the terminal\n",
    "\n",
    " - Now create en env bytyping in the terminal: conda env create -n jupyter-notebook -c anaconda jupyter\n",
    " \n",
    " - To modify the jupyter notebook after downloading the recent copy of the JN from file-zilla: \n",
    " \n",
    " - Activate the env: conda activate jupyter-notebook\n",
    " \n",
    " - Run by typing: jupyter-notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step3**: Genome assembly\n",
    "\n",
    "##########\n",
    "\n",
    "*NOTE:* Before running please make sure you change the IN OUT REF directories paths according to the project.\n",
    "\n",
    "*NOTE:* Before running please make sure you change the the file names in the IN OUT directory to always match *_3_1.fastq.gz and *_3_2.fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (a) **Protegens**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "\n",
    "# Always remember to put all seq in folders named according to the samples\n",
    "# \"/Net/Groups/ccdata/projects/ancientDNA/Plasmid-assembly/Input/XXX_batch/\" \n",
    "# Always remember to put all seq in\n",
    "# \"/Net/Groups/ccdata/projects/ancientDNA/Plasmid-assembly/Input/XXX_batch_ref_seq/\" \n",
    "# before running the script\n",
    "\n",
    "IN=/Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly/Input/p_protegens #change the names accordingly \n",
    "OUT=/Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly/Output/p_protegens #change the names accordingly\n",
    "REF=/Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly/Input/p_protegens_ref_seq #change the names accordingly\n",
    "LP=/Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly/Input/landing_pad\n",
    "\n",
    "mkdir $OUT\n",
    "\n",
    "#############################################################################################\n",
    "#Read filtering\n",
    "#############################################################################################\n",
    "\n",
    "conda activate plasmid_assembly\n",
    "\n",
    "# (1) Run FASTQC on raw data\n",
    "mkdir  $OUT/01_FastQC_results\n",
    "for F in $IN/*; do \n",
    "N=$(basename $F) ;\n",
    "mkdir $OUT/01_FastQC_results/$N ;\n",
    "cd \"$F\"; \n",
    "fastqc $F/*_1.fastq.gz $F/*_2.fastq.gz -t 30 -o $OUT/01_FastQC_results/$N;\n",
    "done \n",
    "\n",
    "# (2) Run trimmomatic\n",
    "mkdir  $OUT/02_Trimmomatic_results \n",
    "for F in $IN/*; do\n",
    "N=$(basename $F) ;\n",
    "mkdir $OUT/02_Trimmomatic_results/$N; \n",
    "cd \"$F\";\n",
    "trimmomatic PE -threads 30 \\\n",
    "-trimlog $N.trimlog.txt \\\n",
    "-summary $N.stats.txt \\\n",
    "$F/*_1.fastq.gz $F/*_2.fastq.gz \\\n",
    "-baseout $N.filtered100.fastq.gz \\\n",
    "MINLEN:100 SLIDINGWINDOW:9:35 ; \n",
    "mv *.txt $OUT/02_Trimmomatic_results/$N\n",
    "mv *_1P.fastq.gz $OUT/02_Trimmomatic_results/$N\n",
    "mv *_2P.fastq.gz $OUT/02_Trimmomatic_results/$N\n",
    "mv *_1U.fastq.gz $OUT/02_Trimmomatic_results/$N\n",
    "mv *_2U.fastq.gz $OUT/02_Trimmomatic_results/$N\n",
    "done \n",
    "\n",
    "# (3) Run FastQC again \n",
    "mkdir $OUT/03_FastQC_results_post_trim\n",
    "for F in $OUT/02_Trimmomatic_results/*; do \n",
    "N=$(basename $F) ;\n",
    "mkdir $OUT/03_FastQC_results_post_trim/$N ;\n",
    "cd \"$F\"; \n",
    "fastqc $F/*_1P.fastq.gz $F/*_2P.fastq.gz -t 30 -o $OUT/03_FastQC_results_post_trim/$N;\n",
    "done\n",
    "\n",
    "#############################################################################################\n",
    "#Denovo genome assembly\n",
    "#############################################################################################\n",
    "\n",
    "conda activate single_genome_assembly\n",
    "\n",
    "mkdir $OUT/06_spades_genome_assembly\n",
    "for F in $OUT/02_Trimmomatic_results/*; do\n",
    "N=$(basename $F) ;\n",
    "mkdir $OUT/06_spades_genome_assembly/$N && cd \"$_\"\n",
    "spades.py --isolate --careful -t 29 \\\n",
    "-1 $OUT/02_Trimmomatic_results/$N/*_1P.fastq.gz \\\n",
    "-2 $OUT/02_Trimmomatic_results/$N/*_2P.fastq.gz \\\n",
    "-o $OUT/06_spades_genome_assembly/$N\n",
    "seqkit fx2tab $OUT/06_spades_genome_assembly/$N/scaffolds.fasta | csvtk mutate -H -t -f 1 -p \"length_([0-9]+)\" | awk -F \"\\t\" '$4>=1000' | seqkit tab2fx > filtered_scaffolds.fasta\n",
    "; done\n",
    "\n",
    "#############################################################################################\n",
    "#Denovo genome assembly::Which Reference genome was it? and where is the LAnding pad?\n",
    "#############################################################################################\n",
    "\n",
    "conda activate plasmid_assembly\n",
    "\n",
    "mkdir  $OUT/05_Minimap2_results_for_denovo5\n",
    "for F in $OUT/05_spades_genome_assembly/*; do \n",
    "N=$(basename $F) ;\n",
    "mkdir $OUT/05_Minimap2_results_for_denovo5/$N && cd \"$_\"\n",
    "minimap2 -ax asm5 $REF/*_dgacA.fasta $OUT/05_spades_genome_assembly/$N/filtered_scaffolds.fasta > aln_ref_genome.sam \n",
    "minimap2 -ax asm5 $LP/*.fna $OUT/05_spades_genome_assembly/$N/filtered_scaffolds.fasta > aln_LP_orig.sam\n",
    "minimap2 -ax asm5 $LP/*.fasta $OUT/05_spades_genome_assembly/$N/filtered_scaffolds.fasta > aln_LP_2BGC_genes.sam\n",
    "#minimap2 -x map-ont -d $LP/originalLP.mmi $LP/*.fna\n",
    "#minimap2 -a $LP/originalLP.mmi $OUT/05_spades_genome_assembly/$N/scaffolds.fasta > test.sam;\n",
    "done\n",
    "\n",
    "\n",
    "#############################################################################################\n",
    "#Denovo genome assembly::Which Reference genome is the genomes?\n",
    "#############################################################################################\n",
    "\n",
    "conda activate taxonomic-binning\n",
    "\n",
    "mkdir $OUT/06_MMseqs2_alignment5\n",
    "for F in $OUT/05_spades_genome_assembly/*; do \n",
    "N=$(basename $F) ;\n",
    "mkdir $OUT/06_MMseqs2_alignment5/$N && cd \"$_\"; \n",
    "mmseqs createdb $REF/*_dgacA.fasta ref\n",
    "#linearise the reference sequence when their is contigs \n",
    "cat $OUT/05_spades_genome_assembly/$N/filtered_scaffolds.fasta | sed -e '1!{/^>.*/d;}' | sed  ':a;N;$!ba;s/\\n//2g' > scaffold.fna\n",
    "mmseqs createdb scaffold.fna scaffold\n",
    "mmseqs prefilter scaffold ref sc_pref \n",
    "mmseqs align scaffold ref sc_pref sc_aln -a --max-seqs 10000\n",
    "mmseqs createtsv scaffold ref sc_aln sc_aln.tsv\n",
    "mmseqs convertalis scaffold ref --format-output \"query,target,evalue,qaln,taln,pident,nident,mismatch\" sc_aln sc_aln.m8 --search-type 3;\n",
    "done\n",
    "\n",
    "#############################################################################################\n",
    "#Ref.based genome alignment\n",
    "#############################################################################################\n",
    "# dgcaA # \n",
    "\n",
    "conda activate plasmid_assembly\n",
    "\n",
    "# Run minimap2 and samtools (reference based alignment) ::: check reference\n",
    "mkdir  $OUT/05_Minimap2_samtools_ref_alignment\n",
    "for F in $OUT/02_Trimmomatic_results/*; do \n",
    "N=$(basename $F) ;\n",
    "mkdir  $OUT/05_Minimap2_samtools_ref_alignment/$N;\n",
    "cd $OUT/05_Minimap2_samtools_ref_alignment/$N;\n",
    "minimap2 -ax sr $REF/*_dgacA.fasta \\\n",
    "$OUT/02_Trimmomatic_results/$N/*_1P.fastq.gz \\\n",
    "$OUT/02_Trimmomatic_results/$N/*_2P.fastq.gz > $N.minimap2.sam; done\n",
    "\n",
    "conda activate samtools\n",
    "\n",
    "for F in $OUT/05_Minimap2_samtools_ref_alignment/*; do \n",
    "N=$(basename $F);\n",
    "cd $OUT/05_Minimap2_samtools_ref_alignment/$N;\n",
    "samtools view -S -b *.sam > aln.bam # Convert sam to bam\n",
    "samtools sort aln.bam -o aln_sorted.bam --threads 28 # Sort the alignment\n",
    "# Get consensus fastq file\n",
    "samtools mpileup -uf $REF/*_dgacA.fasta aln_sorted.bam | bcftools call -c | vcfutils.pl vcf2fq > $N.consensus.fastq\n",
    "# Convert .fastq to .fasta\n",
    "seqtk seq -aQ64 -q13 -n N $N.consensus.fastq > $N.fasta; done\n",
    "\n",
    "\n",
    "# Map the reads to the p.protegens refs (x3)\n",
    "conda activate plasmid_assembly\n",
    "\n",
    "# Run Minimap2 allignment (for all three)\n",
    "mkdir  $OUT/05_Minimap2_results_for_denovo5\n",
    "for F in $OUT/06_spades_genome_assembly/*; do \n",
    "N=$(basename $F) ;\n",
    "mkdir  $OUT/05_Minimap2_results_for_denovo5/$N && cd \"$_\";\n",
    "minimap2 -ax asm20 $REF/*_dgacA.fasta $OUT/06_spades_genome_assembly/$N/contigs.fasta > aln.sam ;\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (a) **Temperata**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "\n",
    "# Always remember to put all seq in folders named according to the samples\n",
    "# \"/Net/Groups/ccdata/projects/ancientDNA/Plasmid-assembly/Input/XXX_batch/\" \n",
    "# Always remember to put all seq in\n",
    "# \"/Net/Groups/ccdata/projects/ancientDNA/Plasmid-assembly/Input/XXX_batch_ref_seq/\" \n",
    "# before running the script\n",
    "\n",
    "IN=/Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly/Input/p_temperata #change the names accordingly \n",
    "OUT=/Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly/Output/p_temperata #change the names accordingly\n",
    "REF=/Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly/Input/p_temperata_ref_seq #change the names accordingly\n",
    "LP=/Net/Groups/ccdata/projects/ancientDNA/Single_genome_assembly/Input/landing_pad\n",
    "\n",
    "mkdir $OUT\n",
    "\n",
    "#############################################################################################\n",
    "#Read filtering\n",
    "#############################################################################################\n",
    "\n",
    "conda activate plasmid_assembly\n",
    "\n",
    "# (1) Run FASTQC on raw data\n",
    "mkdir  $OUT/01_FastQC_results\n",
    "for F in $IN/*; do \n",
    "N=$(basename $F) ;\n",
    "mkdir $OUT/01_FastQC_results/$N ;\n",
    "cd \"$F\"; \n",
    "fastqc $F/*_1.fastq.gz $F/*_2.fastq.gz -t 30 -o $OUT/01_FastQC_results/$N;\n",
    "done \n",
    "\n",
    "# (2) Run trimmomatic\n",
    "mkdir  $OUT/02_Trimmomatic_results \n",
    "for F in $IN/*; do\n",
    "N=$(basename $F) ;\n",
    "mkdir $OUT/02_Trimmomatic_results/$N; \n",
    "cd \"$F\";\n",
    "trimmomatic PE -threads 30 \\\n",
    "-trimlog $N.trimlog.txt \\\n",
    "-summary $N.stats.txt \\\n",
    "$F/*_1.fastq.gz $F/*_2.fastq.gz \\\n",
    "-baseout $N.filtered100.fastq.gz \\\n",
    "MINLEN:100 SLIDINGWINDOW:9:35 ; \n",
    "mv *.txt $OUT/02_Trimmomatic_results/$N\n",
    "mv *_1P.fastq.gz $OUT/02_Trimmomatic_results/$N\n",
    "mv *_2P.fastq.gz $OUT/02_Trimmomatic_results/$N\n",
    "mv *_1U.fastq.gz $OUT/02_Trimmomatic_results/$N\n",
    "mv *_2U.fastq.gz $OUT/02_Trimmomatic_results/$N\n",
    "done \n",
    "\n",
    "# (3) Run FastQC again \n",
    "mkdir $OUT/03_FastQC_results_post_trim\n",
    "for F in $OUT/02_Trimmomatic_results/*; do \n",
    "N=$(basename $F) ;\n",
    "mkdir $OUT/03_FastQC_results_post_trim/$N ;\n",
    "cd \"$F\"; \n",
    "fastqc $F/*_1P.fastq.gz $F/*_2P.fastq.gz -t 30 -o $OUT/03_FastQC_results_post_trim/$N;\n",
    "done\n",
    "\n",
    "#############################################################################################\n",
    "#Denovo genome assembly\n",
    "#############################################################################################\n",
    "\n",
    "conda activate single_genome_assembly\n",
    "\n",
    "mkdir $OUT/05_spades_genome_assembly\n",
    "for F in $OUT/02_Trimmomatic_results/*; do\n",
    "N=$(basename $F) ;\n",
    "mkdir $OUT/05_spades_genome_assembly/$N && cd \"$_\"\n",
    "spades.py -k 21,33,55,77 --only-assembler --careful -t 29 \\\n",
    "-1 $OUT/02_Trimmomatic_results/$N/*_1P.fastq.gz \\\n",
    "-2 $OUT/02_Trimmomatic_results/$N/*_2P.fastq.gz \\\n",
    "-o $OUT/05_spades_genome_assembly/$N\n",
    "seqkit fx2tab $OUT/05_spades_genome_assembly/$N/scaffolds.fasta | csvtk mutate -H -t -f 1 -p \"length_([0-9]+)\" | awk -F \"\\t\" '$4>=1000' | seqkit tab2fx > filtered_scaffolds.fasta;\n",
    "done\n",
    "\n",
    "\n",
    "#############################################################################################\n",
    "#Denovo genome assembly::Which Reference genome is the genomes?\n",
    "#############################################################################################\n",
    "\n",
    "conda activate taxonomic-binning\n",
    "\n",
    "mkdir $OUT/06_MMseqs2_alignment5\n",
    "for F in $OUT/05_spades_genome_assembly/*; do \n",
    "N=$(basename $F) ;\n",
    "mkdir $OUT/06_MMseqs2_alignment5/$N && cd \"$_\";\n",
    "mmseqs createdb $REF/*.fna ref\n",
    "#linearise the reference sequence when their is contigs \n",
    "cat $OUT/05_spades_genome_assembly/$N/filtered_scaffolds.fasta | sed -e '1!{/^>.*/d;}' | sed  ':a;N;$!ba;s/\\n//2g' > scaffold.fna\n",
    "mmseqs createdb scaffold.fna scaffold\n",
    "mmseqs prefilter scaffold ref sc_pref \n",
    "mmseqs align scaffold ref sc_pref sc_aln -a --max-seqs 10000\n",
    "mmseqs createtsv scaffold ref sc_aln sc_aln.tsv\n",
    "mmseqs convertalis scaffold ref --format-output \"query,target,evalue,qaln,taln,pident,nident,mismatch\" sc_aln sc_aln.m8 --search-type 3;\n",
    "done\n",
    "\n",
    "#############################################################################################\n",
    "#Ref.based genome alignment\n",
    "#############################################################################################\n",
    "# dgcaA # \n",
    "\n",
    "conda activate plasmid_assembly\n",
    "\n",
    "# Run minimap2 and samtools (reference based alignment) ::: check reference\n",
    "mkdir  $OUT/05_Minimap2_samtools_ref_alignment\n",
    "for F in $OUT/02_Trimmomatic_results/*; do \n",
    "N=$(basename $F) ;\n",
    "mkdir  $OUT/05_Minimap2_samtools_ref_alignment/$N;\n",
    "cd $OUT/05_Minimap2_samtools_ref_alignment/$N;\n",
    "minimap2 -ax sr $REF/*_dgacA.fasta \\\n",
    "$OUT/02_Trimmomatic_results/$N/*_1P.fastq.gz \\\n",
    "$OUT/02_Trimmomatic_results/$N/*_2P.fastq.gz > $N.minimap2.sam; done\n",
    "\n",
    "conda activate samtools\n",
    "\n",
    "for F in $OUT/05_Minimap2_samtools_ref_alignment/*; do \n",
    "N=$(basename $F);\n",
    "cd $OUT/05_Minimap2_samtools_ref_alignment/$N;\n",
    "samtools view -S -b *.sam > aln.bam # Convert sam to bam\n",
    "samtools sort aln.bam -o aln_sorted.bam --threads 28 # Sort the alignment\n",
    "# Get consensus fastq file\n",
    "samtools mpileup -uf $REF/*_dgacA.fasta aln_sorted.bam | bcftools call -c | vcfutils.pl vcf2fq > $N.consensus.fastq\n",
    "# Convert .fastq to .fasta\n",
    "seqtk seq -aQ64 -q13 -n N $N.consensus.fastq > $N.fasta; done\n",
    "\n",
    "\n",
    "# Map the reads to the p.protegens refs (x3)\n",
    "conda activate plasmid_assembly\n",
    "\n",
    "# Run Minimap2 allignment (for all three)\n",
    "mkdir  $OUT/05_Minimap2_results_for_denovo5\n",
    "for F in $OUT/06_spades_genome_assembly/*; do \n",
    "N=$(basename $F) ;\n",
    "mkdir  $OUT/05_Minimap2_results_for_denovo5/$N && cd \"$_\";\n",
    "minimap2 -ax asm20 $REF/*_dgacA.fasta $OUT/06_spades_genome_assembly/$N/contigs.fasta > aln.sam ;\n",
    "done\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
