{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "50db5bbc",
   "metadata": {},
   "source": [
    "## **Project**: Natural products from the Palaeolithic\n",
    "\n",
    "## **Section**: Genome wide functional analysis: EGGNOG analysis\n",
    "\n",
    "\n",
    "Anan Ibrahim, 01.05.2022\n",
    "\n",
    "**Contents**\n",
    " - **Step1**: Create conda envirorment with required dependencies if not already installed\n",
    " - **Step2**: Download and create the eggnog-database\n",
    " - **Step3**: Prokka annotation of genomes\n",
    " - **Step4**: Map ancient MAGs and modern complete chlorobiaceae genomes to the KEGG database"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b8caa3c",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step1**: Create conda envirorment with required dependencies if not already installed\n",
    "\n",
    "##########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94725496",
   "metadata": {},
   "outputs": [],
   "source": [
    "# All conda envs can be found in EMN001_Paleofuran/02-scripts/ENVS_*.yml\n",
    "conda env create -f prokka.yml\n",
    "conda env create -f eggnog-mapper.yml"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d39733f",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step2**: Download and create the eggnog-database\n",
    "\n",
    "##########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a6639b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir /Net/Groups/ccdata/databases/eggnog-mapper-data\n",
    "cd /Net/Groups/ccdata/apps/conda_envs/eggnog-mapper/lib/python3.8/site-packages/data\n",
    "export EGGNOG_DATA_DIR=/Net/Groups/ccdata/databases/eggnog-mapper-data\n",
    "download_eggnog_data.py --data_dir /Net/Groups/ccdata/databases/eggnog-mapper-data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec08aa78",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step3**: Prokka annotation of genomes\n",
    "\n",
    "##########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a45f943a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############################\n",
    "#Hashes and Directories\n",
    "############################\n",
    "\n",
    "# NOTE: Change directories in bash script accordingly \n",
    "# NOTE: Add the ancient Bins/MAGs in $BINS\n",
    "\n",
    "# Directories: \n",
    "OUT=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Output\n",
    "BINS=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/BINS\n",
    "\n",
    "KEGG_DB=/Net/Groups/ccdata/databases/kegg_db/\n",
    "CHLOROBIALES=/Net/Groups/ccdata/databases/ncbi-ref-genomes/Chlorobiales\n",
    "\n",
    "# Ancient samples:\n",
    "EMN=$BINS/EMN001_021.fna\n",
    "GOY6=$BINS/GOY006_RA.fna\n",
    "GOY5=$BINS/GOY005_001.fna\n",
    "PES=$BINS/PES001_018.fna\n",
    "RIG=$BINS/RIG001_014.fna\n",
    "PLV18=$BINS/PLV001_001.fna\n",
    "PLV20=$BINS/PLV001_002.fna\n",
    "TAF=$BINS/TAF017_RA.fna\n",
    "\n",
    "mkdir $OUT\n",
    "############################\n",
    "# Prokka annotation of MAGs\n",
    "############################\n",
    "mkdir $OUT/PROKKA\n",
    "\n",
    "# Ancient genomes\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate prokka\n",
    "\n",
    "for F in $BINS/*.fna; do \n",
    "  N=$(basename $F .fna) ;\n",
    "  mkdir $OUT/PROKKA/$N ;\n",
    "  prokka --quiet --metagenome --outdir $OUT/PROKKA/$N --prefix $N --locustag $N --force --cpus 0  $F ;\n",
    "done \n",
    "\n",
    "# Modern genomes\n",
    "for F in $CHLOROBIALES/fasta/GCA_013335335.1.fna; do \n",
    "  N=$(basename $F .fna) ;\n",
    "  mkdir $OUT/PROKKA/$N ;\n",
    "  prokka --quiet --metagenome --outdir $OUT/PROKKA/$N --prefix $N --locustag $N --force --cpus 30  $F ;\n",
    "done \n",
    "\n",
    "for F in $CHLOROBIALES/fasta/GCA_013335765.1.fna; do \n",
    "  N=$(basename $F .fna) ;\n",
    "  mkdir $OUT/PROKKA/$N ;\n",
    "  prokka --quiet --metagenome --outdir $OUT/PROKKA/$N --prefix $N --locustag $N --force --cpus 30  $F ;\n",
    "done \n",
    "\n",
    "for F in $CHLOROBIALES/fasta/GCF_000020465.1.fna; do \n",
    "  N=$(basename $F .fna) ;\n",
    "  mkdir $OUT/PROKKA/$N ;\n",
    "  prokka --quiet --metagenome --outdir $OUT/PROKKA/$N --prefix $N --locustag $N --force --cpus 30  $F ;\n",
    "done \n",
    "\n",
    "for F in $CHLOROBIALES/fasta/GCF_001509575.1.1.fna; do \n",
    "  N=$(basename $F .fna) ;\n",
    "  mkdir $OUT/PROKKA/$N ;\n",
    "  prokka --quiet --metagenome --outdir $OUT/PROKKA$N --prefix $N --locustag $N --force --cpus 30  $F ;\n",
    "done \n",
    "\n",
    "for F in $CHLOROBIALES/fasta/GCF_000015125.1.fna; do \n",
    "  N=$(basename $F .fna) ;\n",
    "  mkdir $OUT/PROKKA/$N ;\n",
    "  prokka --quiet --metagenome --outdir $OUT/PROKKA/$N --prefix $N --locustag $N --force --cpus 30  $F ;\n",
    "done \n",
    "\n",
    "for F in $CHLOROBIALES/fasta/GCA_019163275.1.fna; do \n",
    "  N=$(basename $F .fna) ;\n",
    "  mkdir $OUT/PROKKA/$N ;\n",
    "  prokka --quiet --metagenome --outdir $OUT/PROKKA/$N --prefix $N --locustag $N --force --cpus 30  $F ;\n",
    "done \n",
    "\n",
    "conda deactivate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "951eb09b",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step4**: Map ancient MAGs and modern complete chlorobiaceae genomes to the KEGG database\n",
    "\n",
    "##########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfbb7ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############################\n",
    "#Hashes and Directories\n",
    "############################\n",
    "\n",
    "# NOTE: Change directories in bash script accordingly \n",
    "# NOTE: Add the ancient Bins/MAGs in $BINS\n",
    "\n",
    "# Directories: \n",
    "OUT=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Output\n",
    "BINS=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/BINS\n",
    "\n",
    "KEGG_DB=/Net/Groups/ccdata/databases/kegg_db/\n",
    "CHLOROBIALES=/Net/Groups/ccdata/databases/ncbi-ref-genomes/Chlorobiales\n",
    "\n",
    "# Ancient samples:\n",
    "EMN=$BINS/EMN001_021.fna\n",
    "GOY6=$BINS/GOY006_RA.fna\n",
    "GOY5=$BINS/GOY005_001.fna\n",
    "PES=$BINS/PES001_018.fna\n",
    "RIG=$BINS/RIG001_014.fna\n",
    "PLV18=$BINS/PLV001_001.fna\n",
    "PLV20=$BINS/PLV001_002.fna\n",
    "TAF=$BINS/TAF017_RA.fna\n",
    "\n",
    "mkdir $OUT\n",
    "\n",
    "############################\n",
    "# Functional analysis : EGGNOG analysis of the MAGS \n",
    "############################\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate eggnog-mapper\n",
    "\n",
    "mkdir $OUT/EGGNOG\n",
    "cd $OUT/EGGNOG\n",
    "mkdir $OUT/EGGNOG/temp\n",
    "\n",
    "# Ancient bins : Diamond\n",
    "for F in $OUT/PROKKA/*/*.faa; do \n",
    "  N=$(basename $F .faa) ;\n",
    "  mkdir $OUT/EGGNOG/$N ;\n",
    "  emapper.py -i $F -o test --cpu 28 --output_dir $OUT/EGGNOG/$N --temp_dir $OUT/EGGNOG/temp --excel ;\n",
    "  #emapper.py -i $F -o test --cpu 28 --output_dir $OUT/EGGNOG/$N --temp_dir $OUT/EGGNOG/temp --excel --decorate_gff yes ;\n",
    "done\n",
    "\n",
    "# Modern genomes : Diamond\n",
    "for F in $OUT/EGGNOG/*; do\n",
    "sed -i -e '1,4d' $F/test.emapper.annotations;\n",
    "done\n",
    "\n",
    "for F in $OUT/EGGNOG/prokka/GCA_013335335.1/*.faa; do \n",
    "  N=$(basename $F .faa) ;\n",
    "  mkdir $OUT/EGGNOG/$N ;\n",
    "  emapper.py -i $F -o test --cpu 28 --output_dir $OUT/EGGNOG/$N --temp_dir $OUT/EGGNOG/temp --excel ;\n",
    "  sed -i -e '1,4d' $OUT/EGGNOG/$N/test.emapper.annotations\n",
    "done \n",
    "for F in $OUT/EGGNOG/prokka/GCA_013335765.1/*.faa; do \n",
    "  N=$(basename $F .faa) ;\n",
    "  mkdir $OUT/EGGNOG/$N ;\n",
    "  emapper.py -i $F -o test --cpu 28 --output_dir $OUT/EGGNOG/$N --temp_dir $OUT/EGGNOG/temp --excel ;\n",
    "  sed -i -e '1,4d' $OUT/EGGNOG/$N/test.emapper.annotations\n",
    "done\n",
    "for F in $OUT/PROKKA-drep-refgenomes/GCF_000015125.1/*.faa; do \n",
    "  N=$(basename $F .faa) ;\n",
    "  mkdir $OUT/EGGNOG/$N ;\n",
    "  emapper.py -i $F -o test --cpu 28 --output_dir $OUT/EGGNOG/$N --temp_dir $OUT/EGGNOG/temp --excel ;\n",
    "  sed -i -e '1,4d' $OUT/EGGNOG/$N/test.emapper.annotations\n",
    "done \n",
    "for F in $OUT/PROKKA-drep-refgenomes/GCA_019163275.1/*.faa; do \n",
    "  N=$(basename $F .faa) ;\n",
    "  mkdir $OUT/EGGNOG/$N ;\n",
    "  emapper.py -i $F -o test --cpu 28 --output_dir $OUT/EGGNOG/$N --temp_dir $OUT/EGGNOG/temp --excel ;\n",
    "  sed -i -e '1,4d' $OUT/EGGNOG/$N/test.emapper.annotations\n",
    "done \n",
    "for F in $OUT/PROKKA-drep-refgenomes/GCF_001509575.1/*.faa; do \n",
    "  N=$(basename $F .faa) ;\n",
    "  mkdir $OUT/EGGNOG/$N ;\n",
    "  emapper.py -i $F -o test --cpu 28 --output_dir $OUT/EGGNOG/$N --temp_dir $OUT/EGGNOG/temp --excel ;\n",
    "  sed -i -e '1,4d' $OUT/EGGNOG/$N/test.emapper.annotations\n",
    "done \n",
    "for F in $OUT/PROKKA-drep-refgenomes/GCF_000020465.1/*.faa; do \n",
    "  N=$(basename $F .faa) ;\n",
    "  mkdir $OUT/EGGNOG/$N ;\n",
    "  emapper.py -i $F -o test --cpu 28 --output_dir $OUT/EGGNOG/$N --temp_dir $OUT/EGGNOG/temp --excel ;\n",
    "  sed -i -e '1,4d' $OUT/EGGNOG/$N/test.emapper.annotations\n",
    "done \n",
    "for F in $OUT/PROKKA-drep-refgenomes/GCF_001509575.1/*.faa; do \n",
    "  N=$(basename $F .faa) ;\n",
    "  mkdir $OUT/EGGNOG/$N ;\n",
    "  emapper.py -i $F -o test --cpu 28 --output_dir $OUT/EGGNOG/$N --temp_dir $OUT/EGGNOG/temp --excel ;\n",
    "  sed -i -e '1,4d' $OUT/EGGNOG/$N/test.emapper.annotations\n",
    "done \n",
    "\n",
    "conda deactivate\n",
    "\n",
    "# Convert the kegg ids to labels\n",
    "\n",
    "wget -O $KEGG_DB/KO.txt \"http://rest.kegg.jp/list/ko\"\n",
    "wget -O $KEGG_DB/module.txt \"http://rest.kegg.jp/list/module\"\n",
    "wget -O $KEGG_DB/pathway.txt \"http://rest.kegg.jp/list/pathway\"\n",
    "\n",
    "sed -i 's/path://g' $KEGG_DB/pathway.txt\n",
    "echo -e \"KEGG_Pathway\\tLabel\" | cat - $KEGG_DB/pathway.txt > $KEGG_DB/pathway2.txt"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
