{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a26c3090",
   "metadata": {},
   "source": [
    "## **Project**: Natural products from the Palaeolithic\n",
    "\n",
    "## **Section**: Phylogenetic analysis of the dereplicated Chlorobiales genomes\n",
    "\n",
    "\n",
    "Anan Ibrahim, 01.05.2022\n",
    "\n",
    "**Contents**\n",
    " - **Step1**: Create conda envirorment with required dependencies if not already installed\n",
    " - **Step2**: Dereplicate the Chlorobiales genomes\n",
    " - **Step3**: Construct the phylogenetic tree : Chlorobium limicola (aminoacid level)\n",
    " - **Step4**: Construct the phylogenetic tree : Chlorobaculum parvum (aminoacid level)\n",
    " - **Step5**: Construct the phylogenetic tree : Chlorobium limicola (nucleotide level)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23f7f504",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step1**: Create conda envirorment with required dependencies if not already installed\n",
    "\n",
    "##########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a9f59c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# All conda envs can be found in EMN001_Paleofuran/02-scripts/ENVS_*.yml\n",
    "conda env create -f drep.yml\n",
    "conda env create -f phylophlan.yml\n",
    "conda env create -f phyloplan_nucleotide.yml"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d689ab2",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step2**: Dereplicate the Chlorobiales genomes \n",
    "\n",
    "##########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd079ae3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############################\n",
    "#Hashes and Directories\n",
    "############################\n",
    "\n",
    "# NOTE: Change directories in bash script accordingly \n",
    "# NOTE: Add the ancient Bins/MAGs in $BINS\n",
    "\n",
    "# Directories: \n",
    "OUT=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Output\n",
    "BINS=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/BINS\n",
    "\n",
    "CHLOROBIALES=/Net/Groups/ccdata/databases/ncbi-ref-genomes/Chlorobiales\n",
    "DREPGENOME=//Net/Groups/ccdata/databases/ncbi-ref-genomes/Chlorobiales/drep/dereplicated_genomes\n",
    "\n",
    "# Ancient samples:\n",
    "EMN=$BINS/EMN001_021.fna\n",
    "GOY6=$BINS/GOY006_RA.fna\n",
    "GOY5=$BINS/GOY005_001.fna\n",
    "PES=$BINS/PES001_018.fna\n",
    "RIG=$BINS/RIG001_014.fna\n",
    "PLV18=$BINS/PLV001_001.fna\n",
    "PLV20=$BINS/PLV001_002.fna\n",
    "TAF=$BINS/TAF017_RA.fna\n",
    "ROOT=$BINS/JAGYWA010000017.fna\n",
    "\n",
    "mkdir $OUT\n",
    "\n",
    "############################\n",
    "#Drep = dreplicate genomes\n",
    "############################\n",
    "mkdir $CHLOROBIALES/drep\n",
    "\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate drep\n",
    "\n",
    "dRep dereplicate $CHLOROBIALES/drep \\\n",
    "            -p 25 \\\n",
    "            --ignoreGenomeQuality \\\n",
    "            -g $CHLOROBIALES/fasta/*.fna \\\n",
    "            --S_algorithm ANImf \\\n",
    "            -pa 0.90 -sa 0.99\n",
    "\n",
    "cp $BINS/* $DREPGENOME/\n",
    "cp $ROOT $DREPGENOME/\n",
    "\n",
    "conda deactivate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b7d98d8",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step3**: Construct the phylogenetic tree : Chlorobium limicola (aminoacid level)\n",
    "\n",
    "##########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87a00ebd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############################\n",
    "#Hashes and Directories\n",
    "############################\n",
    "\n",
    "# NOTE: Change directories in bash script accordingly \n",
    "# NOTE: Add the ancient Bins/MAGs in $BINS\n",
    "\n",
    "# Directories: \n",
    "OUT=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Output\n",
    "BINS=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/BINS\n",
    "\n",
    "CHLOROBIALES=/Net/Groups/ccdata/databases/ncbi-ref-genomes/Chlorobiales\n",
    "DREPGENOME=//Net/Groups/ccdata/databases/ncbi-ref-genomes/Chlorobiales/drep/dereplicated_genomes\n",
    "\n",
    "# Ancient samples:\n",
    "EMN=$BINS/EMN001_021.fna\n",
    "GOY6=$BINS/GOY006_RA.fna\n",
    "GOY5=$BINS/GOY005_001.fna\n",
    "PES=$BINS/PES001_018.fna\n",
    "RIG=$BINS/RIG001_014.fna\n",
    "PLV18=$BINS/PLV001_001.fna\n",
    "PLV20=$BINS/PLV001_002.fna\n",
    "TAF=$BINS/TAF017_RA.fna\n",
    "ROOT=$BINS/JAGYWA010000017.fna\n",
    "\n",
    "mkdir $OUT\n",
    "\n",
    "############################\n",
    "# PHYLOPHLAN-chlorobium limicola\n",
    "############################\n",
    "\n",
    "mkdir $OUT/TREE_drep_climicola\n",
    "\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate phylophlan\n",
    "\n",
    "cd $OUT/TREE_drep_climicola\n",
    "\n",
    "# Make config file : allignment based on translated sequences : using mafft and diamond. If nucleotide allignents needed use  --force_nucleotides here and in phylophlan command.\n",
    "phylophlan_write_config_file \\\n",
    "    -d a \\\n",
    "    -o  $OUT/TREE_drep_climicola/config.cfg \\\n",
    "    --db_aa diamond \\\n",
    "    --map_dna diamond \\\n",
    "    --map_aa diamond \\\n",
    "    --msa mafft \\\n",
    "    --trim trimal \\\n",
    "    --tree1 fasttree \\\n",
    "    --tree2 raxml\n",
    "    \n",
    "# Set up database (can either just use phylophlan database or specifically chlorobium limicola. \n",
    "# Here chlorobium was used because otherwise the query seqs will be removed due to not fullfilling the hmm models created to retain the sequences.\n",
    "mkdir $OUT/TREE_drep_climicola/core_genes/\n",
    "\n",
    "phylophlan_setup_database \\\n",
    "    -g s__Chlorobium_limicola --database_update \\\n",
    "    -o $OUT/TREE_drep_climicola/core_genes/ \\\n",
    "    --verbose\n",
    "    \n",
    "## Use this flag when phylophlan run crashes  --clean_all \n",
    "phylophlan \\\n",
    "-i $DREPGENOME \\\n",
    "-d $OUT/TREE_drep_climicola/core_genes/s__Chlorobium_limicola \\\n",
    "-o $OUT/TREE_drep_climicola/ \\\n",
    "--diversity medium \\\n",
    "--accurate \\\n",
    "-f $OUT/TREE_drep_climicola/config.cfg \\\n",
    "--nproc 28 --min_num_markers 1 \\\n",
    "--verbose > $OUT/TREE_drep_climicola/phylophlan_output.log\n",
    "\n",
    "############################\n",
    "# Phylophlan alternative with bootstrapping\n",
    "############################\n",
    "\n",
    "cd $OUT/TREE_drep_climicola/\n",
    "raxmlHPC-PTHREADS-SSE3 -m PROTCATLG -f a -x 12345 -p 12345 -# 100 -T 28 -w $OUT/TREE_drep_climicola/ -s $OUT/TREE_drep_climicola/*.aln -n climicola_refined_bootstrap.tre\n",
    "\n",
    "conda deactivate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebd86aa0",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step4**: Construct the phylogenetic tree : Chlorobaculum parvum (aminoacid level)\n",
    "\n",
    "##########\n",
    "\n",
    "*NOTE*: Manually download all chlorobaculum parvum core proteins from Uniref90 https://www.uniprot.org/uniref/?query=taxonomy%3A%22Chlorobaculum+parvum+%5B274539%5D%22+identity%3A0.9&sort=score and add them to (CORECPARVUM)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9741790",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############################\n",
    "#Hashes and Directories\n",
    "############################\n",
    "\n",
    "# NOTE: Change directories in bash script accordingly \n",
    "# NOTE: Add the ancient Bins/MAGs in $BINS\n",
    "\n",
    "# Directories: \n",
    "OUT=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Output\n",
    "BINS=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/BINS\n",
    "\n",
    "CHLOROBIALES=/Net/Groups/ccdata/databases/ncbi-ref-genomes/Chlorobiales\n",
    "DREPGENOME=//Net/Groups/ccdata/databases/ncbi-ref-genomes/Chlorobiales/drep/dereplicated_genomes\n",
    "CORECPARVUM=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/Uniref90/chlorobaculum_parvum\n",
    "\n",
    "# Ancient samples:\n",
    "EMN=$BINS/EMN001_021.fna\n",
    "GOY6=$BINS/GOY006_RA.fna\n",
    "GOY5=$BINS/GOY005_001.fna\n",
    "PES=$BINS/PES001_018.fna\n",
    "RIG=$BINS/RIG001_014.fna\n",
    "PLV18=$BINS/PLV001_001.fna\n",
    "PLV20=$BINS/PLV001_002.fna\n",
    "TAF=$BINS/TAF017_RA.fna\n",
    "ROOT=$BINS/JAGYWA010000017.fna\n",
    "\n",
    "mkdir $OUT\n",
    "\n",
    "############################\n",
    "# PHYLOPHLAN-chlorobaculum parvum\n",
    "############################\n",
    "# NOTE: Manually download all chlorobaculum parvum core proteins from Uniref90 https://www.uniprot.org/uniref/?query=taxonomy%3A%22Chlorobaculum+parvum+%5B274539%5D%22+identity%3A0.9&sort=score and add them to (CORECPARVUM)\n",
    "\n",
    "mkdir $OUT/TREE_drep_cparvum\n",
    "\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate phylophlan\n",
    "\n",
    "cd $OUT/TREE_drep_cparvum\n",
    "\n",
    "# Make config file : allignment based on translated sequences : using mafft and diamond. If nucleotide allignents needed use  --force_nucleotides here and in phylophlan command.\n",
    "phylophlan_write_config_file \\\n",
    "    -d a \\\n",
    "    -o  $OUT/TREE_drep_cparvum/config.cfg \\\n",
    "    --db_aa diamond \\\n",
    "    --map_dna diamond \\\n",
    "    --map_aa diamond \\\n",
    "    --msa mafft \\\n",
    "    --trim trimal \\\n",
    "    --tree1 fasttree \\\n",
    "    --tree2 raxml\n",
    "    \n",
    "# Set up custom database\n",
    "# Manually download all chlorobaculum parvum core proteeins from Uniref90 https://www.uniprot.org/uniref/?query=taxonomy%3A%22Chlorobaculum+parvum+%5B274539%5D%22+identity%3A0.9&sort=score\n",
    "# Add them to : /Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/Uniref90/chlorobaculum_parvum\n",
    "\n",
    "mkdir $OUT/TREE_drep_cparvum/core_genes/\n",
    "phylophlan_setup_database \\\n",
    "    -i $CORECPARVUM/ \\\n",
    "    -d Uniref90_chlorobaculum_parvum \\\n",
    "    -e faa \\\n",
    "    -t a --overwrite -o $OUT/TREE_drep_cparvum/core_genes --verbose\n",
    "    \n",
    "## Use this flag when phylophlan run crashes  --clean_all  also use --databases_folder $OUT/core_genes/ \\\n",
    "phylophlan \\\n",
    "-i $DREPGENOME \\\n",
    "-d $OUT/TREE_drep_cparvum/core_genes/Uniref90_chlorobaculum_parvum \\\n",
    "-o $OUT/TREE_drep_cparvum/ \\\n",
    "--diversity medium \\\n",
    "--accurate \\\n",
    "-f $OUT/TREE_drep_cparvum/config.cfg \\\n",
    "--nproc 30 --min_num_markers 1 \\\n",
    "--verbose > $OUT/TREE_drep_cparvum/phylophlan_output.log\n",
    "\n",
    "############################\n",
    "# Phylophlan alternative with bootstrapping\n",
    "############################\n",
    "cd $OUT/TREE_drep_cparvum/\n",
    "raxmlHPC-PTHREADS-SSE3 -m PROTCATLG -f a -x 12345 -p 12345 -# 100 -T 28 -w $OUT/TREE_drep_cparvum/ -s $OUT/TREE_drep_cparvum/*.aln -n cparvum_refined_bootstrap.tre\n",
    "\n",
    "conda deactivate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb3aa42c",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step5**: Construct the phylogenetic tree : Chlorobium limicola (nucleotide level)\n",
    "\n",
    "##########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92bee899",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############################\n",
    "#Hashes and Directories\n",
    "############################\n",
    "\n",
    "# NOTE: Change directories in bash script accordingly \n",
    "# NOTE: Add the ancient Bins/MAGs in $BINS\n",
    "\n",
    "# Directories: \n",
    "OUT=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Output\n",
    "BINS=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/BINS\n",
    "\n",
    "CHLOROBIALES=/Net/Groups/ccdata/databases/ncbi-ref-genomes/Chlorobiales\n",
    "DREPGENOME=//Net/Groups/ccdata/databases/ncbi-ref-genomes/Chlorobiales/drep/dereplicated_genomes\n",
    "\n",
    "# Ancient samples:\n",
    "EMN=$BINS/EMN001_021.fna\n",
    "GOY6=$BINS/GOY006_RA.fna\n",
    "GOY5=$BINS/GOY005_001.fna\n",
    "PES=$BINS/PES001_018.fna\n",
    "RIG=$BINS/RIG001_014.fna\n",
    "PLV18=$BINS/PLV001_001.fna\n",
    "PLV20=$BINS/PLV001_002.fna\n",
    "TAF=$BINS/TAF017_RA.fna\n",
    "ROOT=$BINS/JAGYWA010000017.fna\n",
    "\n",
    "mkdir $OUT\n",
    "\n",
    "############################\n",
    "# PHYLOPHLAN-chlorobium limicola-drep\n",
    "############################\n",
    "mkdir $OUT/TREE_drep_climicola-nt\n",
    "\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate phyloplan_nucleotide\n",
    "\n",
    "cd $OUT/TREE_drep_climicola-nt\n",
    "\n",
    "# Make config file : allignment based on translated sequences : using mafft and diamond. If nucleotide allignents needed use  --force_nucleotides here and in phylophlan command.\n",
    "phylophlan_write_config_file \\\n",
    "    -d n \\\n",
    "    -o $OUT/TREE_drep_climicola-nt/config.cfg \\\n",
    "    --db_dna makeblastdb \\\n",
    "    --map_dna blastn \\\n",
    "    --msa mafft \\\n",
    "    --trim trimal \\\n",
    "    --tree1 fasttree \\\n",
    "    --tree2 raxml    \n",
    "    \n",
    "# Set up database (can either just use phylophlan database or specifically chlorobium limicola. \n",
    "# Here chlorobium was used because otherwise the query seqs will be removed due to not fullfilling the hmm models created to retain the sequences.\n",
    "mkdir $OUT/TREE_drep_climicola-nt/core_genes/\n",
    "\n",
    "phylophlan_setup_database \\\n",
    "    -g s__Chlorobium_limicola -t n --database_update \\\n",
    "    -o $OUT/TREE_drep_climicola-nt/core_genes/ \\\n",
    "    --verbose\n",
    "    \n",
    "## Use this flag when phylophlan run crashes  --clean_all \n",
    "phylophlan \\\n",
    "-i $DREPGENOME \\\n",
    "-d $OUT/TREE_drep_climicola-nt/core_genes/s__Chlorobium_limicola \\\n",
    "-o $OUT/TREE_drep_climicola-nt/ \\\n",
    "--diversity medium \\\n",
    "--accurate \\\n",
    "-f $OUT/TREE_drep_climicola-nt/config.cfg \\\n",
    "--nproc 28 --min_num_markers 1 \\\n",
    "--verbose > $OUT/TREE_drep_climicola-nt/phylophlan_output.log\n",
    "\n",
    "#THE ROOT was not incoportaed into the tree because it did not contain any similar core genes as the chlorobiales genomes\n",
    "\n",
    "############################\n",
    "# Phylophlan alternative with bootstrapping\n",
    "############################\n",
    "cd $OUT/TREE_drep_climicola-nt/\n",
    "raxmlHPC-PTHREADS-SSE3 -m GTRCAT -f a -x 12345 -p 12345 -# 100 -T 28 -w $OUT/TREE_drep_climicola-nt/ -s $OUT/TREE_drep_climicola-nt/*.aln -n climicola_nt_refined_bootstrap.tre\n",
    "\n",
    "conda deactivate"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
