{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "51776a5d",
   "metadata": {},
   "source": [
    "## **Project**: Natural products from the Palaeolithic\n",
    "\n",
    "## **Section**: Download the entire Chlorobiales genomes and construct a phylogenetic tree\n",
    "\n",
    "\n",
    "Anan Ibrahim, 01.05.2022\n",
    "\n",
    "**Contents**\n",
    " - **Step1**: Create conda envirorment with required dependencies if not already installed \n",
    " - **Step2**: Download Chlorobiales genomes from the Assembly database\n",
    " - **Step3**: Construct entire Chlorobiales order + Ancient MAGs phylogenetic tree\n",
    " - **Step4**: Prune the tree to only include the clade with ancient MAGs\n",
    " - **Step5**: Calculate the ANI values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "032a3b5b",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step1**: Create conda envirorment with required dependencies if not already installed \n",
    "\n",
    "##########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7cc3edf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# All conda envs can be found in EMN001_Paleofuran/02-scripts/ENVS_*.yml\n",
    "conda env create -f chlorobiales_genomes1.yml\n",
    "conda env create -f chlorobiales_genomes2.yml\n",
    "conda env create -f phylophlan.yml\n",
    "conda env create -f fastani.yml"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c63116dd",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step2**: Download Chlorobiales genomes from the Assembly database\n",
    "\n",
    "##########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b95b0f6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############################\n",
    "#Hashes and Directories\n",
    "############################\n",
    "\n",
    "# NOTE: Change directories in bash script accordingly \n",
    "# NOTE: Add the ancient Bins/MAGs in $BINS\n",
    "\n",
    "# Directories: \n",
    "OUT=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Output\n",
    "BINS=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/BINS\n",
    "\n",
    "CHLOROBIALES=/Net/Groups/ccdata/databases/ncbi-ref-genomes/Chlorobiales\n",
    "\n",
    "# Ancient samples:\n",
    "EMN=$BINS/EMN001_021.fna\n",
    "GOY6=$BINS/GOY006_RA.fna\n",
    "GOY5=$BINS/GOY005_001.fna\n",
    "PES=$BINS/PES001_018.fna\n",
    "RIG=$BINS/RIG001_014.fna\n",
    "PLV18=$BINS/PLV001_001.fna\n",
    "PLV20=$BINS/PLV001_002.fna\n",
    "TAF=$BINS/TAF017_RA.fna\n",
    "\n",
    "mkdir $OUT\n",
    "\n",
    "############################\n",
    "#Download from Assembly (NCBI)\n",
    "############################\n",
    "mkdir $CHLOROBIALES\n",
    "mkdir $CHLOROBIALES/fasta\n",
    "\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate chlorobiales_genomes1\n",
    "\n",
    "cd $CHLOROBIALES\n",
    "\n",
    "# Grab the search query from the Assembly database on ncbi first  \n",
    "esearch -db assembly -query '\"Chlorobiales\"[Organism] AND (latest[filter] AND all[filter] NOT anomalous[filter])' \\\n",
    "| esummary | xtract -pattern DocumentSummary -element AssemblyAccession > chlorobiales_assembly-accs.txt\n",
    "#Result: 599 seq. \n",
    "conda deactivate\n",
    "\n",
    "conda activate chlorobiales_genomes2\n",
    "\n",
    "# Download the fasta files corresponnding to the accessions\n",
    "# -j parallelism \n",
    "cd $CHLOROBIALES/fasta\n",
    "\n",
    "bit-dl-ncbi-assemblies -w $CHLOROBIALES/chlorobiales_assembly-accs.txt -f fasta -j 20\n",
    "# Result: 596 genomes downloaded (the rest not found)\n",
    "gzip -d *.gz \n",
    "\n",
    "conda deactivate \n",
    "\n",
    "cp $BINS/* $CHLOROBIALES/fasta\n",
    "cp $ROOT $CHLOROBIALES/fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4506a632",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step3**: Construct entire Chlorobiales order + Ancient MAGs phylogenetic tree\n",
    "\n",
    "##########\n",
    "\n",
    "*Note:* Do not forget to add the outgroup in this group we downloaded Flavobacterium branchiicola seperately from NCBI:@https://www.ncbi.nlm.nih.gov/nuccore/NZ_JAGYWA010000017"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15f8a395",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############################\n",
    "#Hashes and Directories\n",
    "############################\n",
    "\n",
    "# NOTE: Change directories in bash script accordingly \n",
    "# NOTE: Add the ancient Bins in $BINS\n",
    "\n",
    "# Directories: \n",
    "OUT=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Output\n",
    "BINS=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/BINS\n",
    "\n",
    "CHLOROBIALES=/Net/Groups/ccdata/databases/ncbi-ref-genomes/Chlorobiales\n",
    "\n",
    "# Samples:\n",
    "EMN=$BINS/EMN001_021.fna\n",
    "GOY6=$BINS/GOY006_RA.fna\n",
    "GOY5=$BINS/GOY005_001.fna\n",
    "PES=$BINS/PES001_018.fna\n",
    "RIG=$BINS/RIG001_014.fna\n",
    "PLV18=$BINS/PLV001_001.fna\n",
    "PLV20=$BINS/PLV001_002.fna\n",
    "TAF=$BINS/TAF017_RA.fna\n",
    "ROOT=$BINS/JAGYWA010000017.fna\n",
    "\n",
    "mkdir $OUT\n",
    "\n",
    "############################\n",
    "# PHYLOPHLAN-chlorobium limicola-all 596 genomes\n",
    "############################\n",
    "mkdir $OUT/TREE_chlorobiales_genomes\n",
    "\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate phylophlan\n",
    "\n",
    "cd $OUT/TREE_chlorobiales_genomes\n",
    "\n",
    "# Make config file : allignment based on translated sequences : using mafft and diamond. If nucleotide allignents needed use  --force_nucleotides here and in phylophlan command.\n",
    "phylophlan_write_config_file \\\n",
    "    -d a \\\n",
    "    -o $OUT/TREE_chlorobiales_genomes/config.cfg \\\n",
    "    --db_aa diamond \\\n",
    "    --map_dna diamond \\\n",
    "    --map_aa diamond \\\n",
    "    --msa mafft \\\n",
    "    --trim trimal \\\n",
    "    --tree1 fasttree \\\n",
    "    --tree2 raxml\n",
    "    \n",
    "# Set up database (can either just use phylophlan database or specifically chlorobium limicola. \n",
    "# Here chlorobium was used because otherwise the query seqs will be removed due to not fullfilling the hmm models created to retain the sequences.\n",
    "mkdir $OUT/TREE_chlorobiales_genomes/core_genes/\n",
    "\n",
    "phylophlan_setup_database \\\n",
    "    -g s__Chlorobium_limicola --database_update \\\n",
    "    -o $OUT/TREE_chlorobiales_genomes/core_genes/ \\\n",
    "    --verbose\n",
    "    \n",
    "## Use this flag when phylophlan run crashes  --clean_all \n",
    "phylophlan \\\n",
    "-i $CHLOROBIALES/fasta \\\n",
    "-d $OUT/TREE_chlorobiales_genomes/core_genes/s__Chlorobium_limicola \\\n",
    "-o $OUT/TREE_chlorobiales_genomes/ \\\n",
    "--diversity medium \\\n",
    "--accurate \\\n",
    "-f $OUT/TREE_chlorobiales_genomes/config.cfg \\\n",
    "--nproc 30 --min_num_markers 1 \\\n",
    "--verbose > $OUT/TREE_chlorobiales_genomes/phylophlan_output.log\n",
    "\n",
    "# Bootstrap the tree\n",
    "cd $OUT/TREE_chlorobiales_genomes/\n",
    "raxmlHPC-PTHREADS-SSE3 -s fasta_concatenated.aln -n bootstrap_tree.tre -f a -m PROTGAMMAAUTO -N 100 -x 12345 -p 12345 -T 30\n",
    "\n",
    "conda deactivate\n",
    "\n",
    "##############################################\n",
    "# ANNOTATE THE TREE\n",
    "##############################################\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate chlorobiales_genomes1\n",
    "\n",
    "cd $OUT/TREE_chlorobiales_genomes\n",
    "\n",
    "# Grab the assembly accession for annotation\n",
    "ls $CHLOROBIALES/fasta > chlorobiales_assembly-accs.txt\n",
    "# ls /Net/Groups/ccdata/databases/ncbi-ref-genomes/pseudomonas_syringae/fasta > ass_acc.txt\n",
    "sed -i 's/ /_/g'  chlorobiales_assembly-accs.txt\n",
    "sed -i 's/.fna//g' chlorobiales_assembly-accs.txt\n",
    "\n",
    "file=chlorobiales_assembly-accs.txt\n",
    "lines=$(cat $file)\n",
    "\n",
    "for line in $lines\n",
    "do\n",
    "esearch -db assembly -query \"$line\" |esummary | \\\n",
    "xtract -pattern DocumentSummary \\\n",
    "-sep \"\\t\" -element AssemblyAccession,Taxid,AssemblyName,Organism,assembly-status >> ass_acc_2.txt\n",
    "done\n",
    "\n",
    "# Create a iTOL accession to lineage dataset:\n",
    "# Print the acc. list\n",
    "awk -F '\\t' '{print $1\"\\t\"$4}' \\\n",
    "ass_acc_2.txt > ass_acc_itol2.txt\n",
    "\n",
    "# Manually :\n",
    "# Using this file just open the constructed tree in Itol and drag this file in it to label the branches.\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4d4ea86",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step4**: Prune the tree to only include the clade with ancient MAGs\n",
    "\n",
    "##########\n",
    "\n",
    "*Note:* Do not forget to add the outgroup in this group we downloaded Flavobacterium branchiicola (NZ_JAGYWA010000017) seperately from NCBI\n",
    "\n",
    "*Note:* Visualise the entire tree in the ITOL desktop app, and then grab the acc numbers of the pruned tree that contains all genomes classified upto a species level to a text file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "954cc5c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############################\n",
    "#Hashes and Directories\n",
    "############################\n",
    "\n",
    "# NOTE: Change directories in bash script accordingly \n",
    "# NOTE: Add the ancient Bins in $BINS\n",
    "\n",
    "# Directories: \n",
    "OUT=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Output\n",
    "BINS=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/BINS\n",
    "\n",
    "CHLOROBIALES=/Net/Groups/ccdata/databases/ncbi-ref-genomes/Chlorobiales\n",
    "\n",
    "# Samples:\n",
    "EMN=$BINS/EMN001_021.fna\n",
    "GOY6=$BINS/GOY006_RA.fna\n",
    "GOY5=$BINS/GOY005_001.fna\n",
    "PES=$BINS/PES001_018.fna\n",
    "RIG=$BINS/RIG001_014.fna\n",
    "PLV18=$BINS/PLV001_001.fna\n",
    "PLV20=$BINS/PLV001_002.fna\n",
    "TAF=$BINS/TAF017_RA.fna\n",
    "ROOT=$BINS/JAGYWA010000017.fna\n",
    "\n",
    "mkdir $OUT\n",
    "\n",
    "############################\n",
    "# Ancient Clade: pruned assembly accs and tree\n",
    "############################\n",
    "\n",
    "mkdir $OUT/TREE_chlorobiales_genomes/pruned2\n",
    "\n",
    "# MANUALLY: from itol a text file with the accessions (of spp identified upto a sp. level is created == PHYL_chlorobiales_genomes_annotated_to_spp.txt)\n",
    "# Copy the pruned accessions to another folder and add underscores and .fna to the end of the files \n",
    "sed -i 's/ /_/g' $OUT/TREE_chlorobiales_genomes/pruned2/PHYL_chlorobiales_genomes_annotated_to_spp.txt\n",
    "sed -i \"s|$|.fna|\" $OUT/TREE_chlorobiales_genomes/pruned2/PHYL_chlorobiales_genomes_annotated_to_spp.txt\n",
    "\n",
    "# copy the 31 genomes to another input directory\n",
    "cd $CHLOROBIALES/fasta\n",
    "for file in $(cat $OUT/TREE_chlorobiales_genomes/pruned2/PHYL_chlorobiales_genomes_annotated_to_spp.txt); do \n",
    "cp \"$file\" $OUT/TREE_chlorobiales_genomes/pruned2/; done\n",
    "\n",
    "# Make a tree\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate phylophlan\n",
    "\n",
    "cd $OUT/TREE_chlorobiales_genomes/pruned2\n",
    "\n",
    "# Make config file : allignment based on translated sequences : using mafft and diamond. If nucleotide allignents needed use  --force_nucleotides here and in phylophlan command.\n",
    "phylophlan_write_config_file \\\n",
    "    -d a \\\n",
    "    -o $OUT/TREE_chlorobiales_genomes/pruned2/config.cfg \\\n",
    "    --db_aa diamond \\\n",
    "    --map_dna diamond \\\n",
    "    --map_aa diamond \\\n",
    "    --msa mafft \\\n",
    "    --trim trimal \\\n",
    "    --tree1 fasttree \\\n",
    "    --tree2 raxml\n",
    "    \n",
    "# Set up database (can either just use phylophlan database or specifically chlorobium limicola. \n",
    "# Here chlorobium was used because otherwise the query seqs will be removed due to not fullfilling the hmm models created to retain the sequences.\n",
    "phylophlan_setup_database \\\n",
    "    -g s__Chlorobium_limicola --database_update \\\n",
    "    -o $OUT/TREE_chlorobiales_genomes/core_genes/ \\\n",
    "    --verbose\n",
    "    \n",
    "## Use this flag when phylophlan run crashes  --clean_all \n",
    "phylophlan \\\n",
    "-i $OUT/TREE_chlorobiales_genomes/pruned2 \\\n",
    "-d $OUT/TREE_chlorobiales_genomes/core_genes/s__Chlorobium_limicola \\\n",
    "-o $OUT/TREE_chlorobiales_genomes/pruned2/ \\\n",
    "--diversity low \\\n",
    "--accurate \\\n",
    "-f $OUT/TREE_chlorobiales_genomes/pruned2/config.cfg \\\n",
    "--nproc 30 --min_num_markers 1 \\\n",
    "--verbose > $OUT/TREE_chlorobiales_genomes/pruned2/phylophlan_output.log\n",
    "\n",
    "############################\n",
    "# Phylophlan alternative with bootstrapping \n",
    "############################\n",
    "cd $OUT/TREE_chlorobiales_genomes/pruned2\n",
    "raxmlHPC-PTHREADS-SSE3 -m PROTCATLG -f a -x 12345 -p 12345 -# 100 -T 28 -w $OUT/TREE_chlorobiales_genomes/pruned2 -s $OUT/TREE_chlorobiales_genomes/pruned2/*.aln -n pruned2_refined_bootstrap.tre\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f86259be",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step5**: Calculate the ANI values \n",
    "\n",
    "##########\n",
    "\n",
    "*Note:* Do not forget to add the outgroup in this group we downloaded Flavobacterium branchiicola (NZ_JAGYWA010000017) seperately from NCBI\n",
    "\n",
    "*Note:* Visualise the entire tree in the ITOL desktop app, and then grab the acc numbers of the pruned tree that contains all genomes classified upto a species level to a text file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82d2d477",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############################\n",
    "#Hashes and Directories\n",
    "############################\n",
    "\n",
    "# NOTE: Change directories in bash script accordingly \n",
    "# NOTE: Add the ancient Bins in $BINS\n",
    "\n",
    "# Directories: \n",
    "OUT=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Output\n",
    "BINS=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/BINS\n",
    "\n",
    "CHLOROBIALES=/Net/Groups/ccdata/databases/ncbi-ref-genomes/Chlorobiales\n",
    "\n",
    "# Samples:\n",
    "EMN=$BINS/EMN001_021.fna\n",
    "GOY6=$BINS/GOY006_RA.fna\n",
    "GOY5=$BINS/GOY005_001.fna\n",
    "PES=$BINS/PES001_018.fna\n",
    "RIG=$BINS/RIG001_014.fna\n",
    "PLV18=$BINS/PLV001_001.fna\n",
    "PLV20=$BINS/PLV001_002.fna\n",
    "TAF=$BINS/TAF017_RA.fna\n",
    "ROOT=$BINS/JAGYWA010000017.fna\n",
    "\n",
    "mkdir $OUT\n",
    "\n",
    "############################\n",
    "# Ancient Clade: add the ANI values\n",
    "############################\n",
    "\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate fastani\n",
    "\n",
    "# Calculate ANI\n",
    "cd $OUT/TREE_chlorobiales_genomes/pruned2\n",
    "# conda create -n fastani -c conda-forge -c bioconda fastani\n",
    "\n",
    "# add paths to every line\n",
    "#cp $OUT/TREE_chlorobiales_genomes/pruned2/accessions_IDs_annotated_to_species_level.txt $OUT/TREE_chlorobiales_genomes/pruned2/acc-list-paths.txt \n",
    "#awk '$0=\"$OUT/TREE_chlorobiales_genomes/pruned2/\"$0' $OUT/TREE_chlorobiales_genomes/pruned2/acc-list-paths.txt > acc-list-paths-written.txt\n",
    "\n",
    "#fragLen 3000\n",
    "fastANI --ql accessions_IDs_annotated_to_species_level.txt --rl accessions_IDs_annotated_to_species_level.txt -t 8 -k 16 --fragLen 3000 -o ANI_output_fg3000.txt --matrix\n",
    "#For heatmap ANI estimate figure wo filtering \n",
    "awk -F '\\t' '{print $1\"\\t\"$2\"\\t\"$3}' ANI_output_fg3000.txt > ANI_output1_fg3000.txt\n",
    "\n",
    "awk 'BEGIN {FS=OFS=\"\\t\"} \n",
    "           {col[$1]; row[$2]; val[$2,$1]=$3}\n",
    "     END   {for(c in col) printf \"%s\", OFS c; print \"\";\n",
    "            for(r in row)\n",
    "              {printf \"%s\", r;\n",
    "               for(c in col) printf \"%s\", OFS val[r,c]\n",
    "               print \"\"}}' ANI_output1_fg3000.txt > ANI_OUT_MATRIX_fg3000.txt\n",
    "\n",
    "#fragLen 1000\n",
    "fastANI --ql accessions_IDs_annotated_to_species_level.txt --rl accessions_IDs_annotated_to_species_level.txt -t 8 -k 16 --fragLen 1000 -o ANI_output_fg1000.txt --matrix\n",
    "#For heatmap ANI estimate figure wo filtering \n",
    "awk -F '\\t' '{print $1\"\\t\"$2\"\\t\"$3}' ANI_output_fg1000.txt > ANI_output1_fg1000.txt\n",
    "awk 'BEGIN {FS=OFS=\"\\t\"} \n",
    "           {col[$1]; row[$2]; val[$2,$1]=$3}\n",
    "     END   {for(c in col) printf \"%s\", OFS c; print \"\";\n",
    "            for(r in row)\n",
    "              {printf \"%s\", r;\n",
    "               for(c in col) printf \"%s\", OFS val[r,c]\n",
    "               print \"\"}}' ANI_output1_fg1000.txt > ANI_OUT_MATRIX_fg1000.txt\n",
    "\n",
    "#For heatmap ANI estimate figure w filtering : remove all mappings that had a mapping percentage of less than 70%\n",
    "# fragLen 3000\n",
    "awk -F '\\t' '{print $1\"\\t\"$2\"\\t\"$3\"\\t\"$4\"\\t\"$5\"\\t\"$4/$5}' ANI_output_fg3000.txt > ANI_output2_fg3000.txt\n",
    "awk '$6 >= 0.60 { print }' ANI_output2_fg3000.txt > ANI_output2_fg3000_filter.txt\n",
    "awk '{print $1\"\\t\"$2\"\\t\"$3}' ANI_output2_fg3000_filter.txt > ANI_output21_fg3000_filter.txt\n",
    "awk 'BEGIN {FS=OFS=\"\\t\"} \n",
    "           {col[$1]; row[$2]; val[$2,$1]=$3}\n",
    "     END   {for(c in col) printf \"%s\", OFS c; print \"\";\n",
    "            for(r in row)\n",
    "              {printf \"%s\", r;\n",
    "               for(c in col) printf \"%s\", OFS val[r,c]\n",
    "               print \"\"}}' ANI_output21_fg3000_filter.txt > ANI_OUT_MATRIX_fg3000_filter.txt\n",
    "# fragLen 1000\n",
    "awk -F '\\t' '{print $1\"\\t\"$2\"\\t\"$3\"\\t\"$4\"\\t\"$5\"\\t\"$4/$5}' ANI_output_fg1000.txt > ANI_output2_fg1000.txt\n",
    "awk '$6 >= 0.60 { print }' ANI_output2_fg1000.txt > ANI_output2_fg1000_filter.txt\n",
    "awk '{print $1\"\\t\"$2\"\\t\"$3}' ANI_output2_fg1000_filter.txt > ANI_output21_fg1000_filter.txt\n",
    "awk 'BEGIN {FS=OFS=\"\\t\"} \n",
    "           {col[$1]; row[$2]; val[$2,$1]=$3}\n",
    "     END   {for(c in col) printf \"%s\", OFS c; print \"\";\n",
    "            for(r in row)\n",
    "              {printf \"%s\", r;\n",
    "               for(c in col) printf \"%s\", OFS val[r,c]\n",
    "               print \"\"}}' ANI_output21_fg1000_filter.txt > ANI_OUT_MATRIX_fg1000_filter.txt\n",
    "\n",
    "############################\n",
    "# Ancient Clade: add the source of bacteria\n",
    "############################\n",
    "# This was done manually using the NCBI and checking the publications individually"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72ef6c31",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
