{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6e5d7004",
   "metadata": {},
   "source": [
    "## **Project**: Natural products from the Palaeolithic\n",
    "\n",
    "## **Section**: Phylogenetic analysis of the Butyrolactone BGCs\n",
    "\n",
    "\n",
    "Anan Ibrahim, 01.05.2022\n",
    "\n",
    "**Contents**\n",
    " - **Step1**: Create conda envirorment with required dependencies if not already installed\n",
    " - **Step2**: Phylogenetic network analysis - BiGSCAPE\n",
    " - **Step3**: BGC genetic topolgy - CORASON\n",
    " - **Step4**: AFSA family phylogenetic tree\n",
    " - **Step5**: Alignment of the core genes of the Butyrolactone BGC\n",
    " - **Step6**: Butyrolactone cluster synteny - CLINKER"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "290c568e",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step1**: Create conda envirorment with required dependencies if not already installed\n",
    "\n",
    "##########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "052c5ea5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# All conda envs can be found in EMN001_Paleofuran/02-scripts/ENVS_*.yml\n",
    "conda env create -f chlorobiales_genomes1.yml\n",
    "conda env create -f phylophlan.yml\n",
    "conda env create -f taxonomic-binning.yml\n",
    "conda env create -f clinker.yml"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7d21fcb",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step2**: Phylogenetic network analysis - BiGSCAPE\n",
    "\n",
    "##########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ddb3e06",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############################\n",
    "#Hashes and Directories\n",
    "############################\n",
    "\n",
    "# NOTE: Change directories in bash script accordingly \n",
    "# NOTE: Add the ancient Bins/MAGs in $BINS\n",
    "\n",
    "# Directories: \n",
    "OUT=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Output\n",
    "BINS=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/BINS\n",
    "\n",
    "DREPGENOME=/Net/Groups/ccdata/databases/ncbi-ref-genomes/Chlorobiales/drep/dereplicated_genomes\n",
    "ANCIENT_CONTIGS=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/ancient_contigs_names_butyrolactone\n",
    "\n",
    "mkdir $OUT\n",
    "\n",
    "############################\n",
    "# Bigscape annotation\n",
    "############################\n",
    "mkdir $OUT/BIGSCAPE\n",
    "mkdir $OUT/BIGSCAPE/gbk_regions\n",
    "mkdir $OUT/BIGSCAPE/genome_gbks\n",
    "\n",
    "cd $OUT/BIGSCAPE\n",
    "git clone https://git.wur.nl/medema-group/BiG-SCAPE.git\n",
    "cd $OUT/BIGSCAPE/BiG-SCAPE\n",
    "\n",
    "for S in $OUT/ANTISMASH/* ; do\n",
    "cp $S/*.region001.gbk $OUT/BIGSCAPE/gbk_regions;\n",
    "cp $S/*.region002.gbk $OUT/BIGSCAPE/gbk_regions;\n",
    "cp $S/*.region003.gbk $OUT/BIGSCAPE/gbk_regions;\n",
    "cp $S/*.region004.gbk $OUT/BIGSCAPE/gbk_regions;\n",
    "cp $S/*.region005.gbk $OUT/BIGSCAPE/gbk_regions;\n",
    "done\n",
    "\n",
    "for S in $OUT/ANTISMASH-drep-refgenomes/* ; do\n",
    "cp $S/*.region001.gbk $OUT/BIGSCAPE/gbk_regions;\n",
    "cp $S/*.region002.gbk $OUT/BIGSCAPE/gbk_regions;\n",
    "cp $S/*.region003.gbk $OUT/BIGSCAPE/gbk_regions;\n",
    "cp $S/*.region004.gbk $OUT/BIGSCAPE/gbk_regions;\n",
    "cp $S/*.region005.gbk $OUT/BIGSCAPE/gbk_regions;\n",
    "cp $S/*.region006.gbk $OUT/BIGSCAPE/gbk_regions;\n",
    "cp $S/*.region007.gbk $OUT/BIGSCAPE/gbk_regions;\n",
    "done\n",
    "\n",
    "cp $OUT/PROKKA/*/*.gbk $OUT/BIGSCAPE/genome_gbks\n",
    "cp $OUT/PROKKA-drep-refgenomes/*/*.gbk $OUT/BIGSCAPE/genome_gbks\n",
    "\n",
    "mkdir $OUT/BIGSCAPE/BUTYR-1\n",
    "python bigscape.py --label chlorobiales-drep-butyrolactone --inputdir $OUT/BIGSCAPE/gbk_regions \\\n",
    "--outputdir $OUT/BIGSCAPE/BUTYR-1 \\\n",
    "--pfam_dir /Net/Groups/ccdata/databases/pfam34/ \\\n",
    "--force_hmmscan --mode auto --cutoffs 1.0 --clan_cutoff 1.0 1.0 --include_singletons \\\n",
    "--query_bgc $OUT/ANTISMASH/EMN001_021/c00006_EMN001_...region001.gbk \\\n",
    "--cores 29 --verbose > $OUT/BIGSCAPE/BUTYR-1/run.log\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74ebe82c",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step3**: BGC genetic topolgy - CORASON\n",
    "\n",
    "##########\n",
    "\n",
    "*Note*: Requires Docker installation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fabeec0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#1 mkdir ~/bin\n",
    "#2 pull the docker container: docker pull nselem/corason:latest\n",
    "#3 hcange the windows path in the script : /home/aibrahim/bin\n",
    "#4 download the script: curl -q https://raw.githubusercontent.com/nselem/corason/master/run_corason > ~/bin/run_corason\n",
    "#5 execute the script: chmod a+x ~/bin/run_corason\n",
    "#6 check if its running: ~/bin/run_corason\n",
    "\n",
    "# The <EMN001_AFSA.faa> is the aminoacid sequence of teh AFSA\n",
    "# The <genome_gbks/EMN001_021.gbk> are the gbk files of the ref modern genomes and the ancient MAGs generated by PROKKA\n",
    "sudo ./run_corason EMN001_AFSA.faa genome_gbks genome_gbks/EMN001_021.gbk -g 10"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7480c631",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step4**: AFSA family phylogenetic tree\n",
    "\n",
    "##########\n",
    "\n",
    "*Note*: Manually download all AFSA family proteins (same domain architecture) from InterPro http://www.ebi.ac.uk/interpro/entry/InterPro/IPR005509/protein/UniProt/#table and add them to (AFSAFAMILY)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79c9b5f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############################\n",
    "#Hashes and Directories\n",
    "############################\n",
    "\n",
    "# NOTE: Change directories in bash script accordingly\n",
    "\n",
    "# Directories: \n",
    "OUT=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Output\n",
    "BINS=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/BINS\n",
    "SCRIPT=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Scripts\n",
    "\n",
    "AFSAFAMILY=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/Interpro_A0A101JTJ5\n",
    "ROOT_AFSA=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/AFSA_outgroup\n",
    "ANCIENT_AFSA=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/AFSA_aa\n",
    "\n",
    "# Samples:\n",
    "EMN=$BINS/EMN001_021.fna\n",
    "GOY6=$BINS/GOY006_RA.fna\n",
    "GOY5=$BINS/GOY005_001.fna\n",
    "PES=$BINS/PES001_018.fna\n",
    "RIG=$BINS/RIG001_014.fna\n",
    "PLV18=$BINS/PLV001_001.fna\n",
    "PLV20=$BINS/PLV001_002.fna\n",
    "TAF=$BINS/TAF017_RA.fna\n",
    "ROOT=$BINS/JAGYWA010000017.fna\n",
    "\n",
    "############################\n",
    "#Cluster the AFSA protein family\n",
    "############################\n",
    "# Cluster the 742 AFSA sequences according to 0.5% identity for an accurate phylogenetic tree\n",
    "mkdir $OUT/AFSA_family_cluster\n",
    "mkdir $OUT/AFSA_family_cluster/mmseqs/\n",
    "mkdir $OUT/AFSA_family_cluster/mmseqs/tmp\n",
    "\n",
    "cd $OUT/AFSA_family_cluster/mmseqs\n",
    "\n",
    "# Replace space with underscores\n",
    "sed 's/ /_/g' -i $AFSAFAMILY/*.faa\n",
    "\n",
    "# MMSeqs2 = Cluster the 742 protein sequences\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate taxonomic-binning\n",
    "\n",
    "# create db\n",
    "mmseqs createdb $AFSAFAMILY/*.faa afsa_database\n",
    "\n",
    "# cluster the db -s 7 (sensitivty)\n",
    "mmseqs cluster afsa_database afsa_database_clu tmp --cluster-mode 0 --min-seq-id 0.5 --threads 30\n",
    "\n",
    "# create a table version\n",
    "mmseqs createtsv afsa_database afsa_database afsa_database_clu afsa_database_clu.tsv\n",
    "\n",
    "# create a fasta version\n",
    "mmseqs createseqfiledb afsa_database afsa_database_clu afsa_database_clu_seq \n",
    "mmseqs result2flat afsa_database afsa_database afsa_database_clu_seq afsa_database_clu_seq.fasta\n",
    "\n",
    "# extract the representative of a clustering \n",
    "mmseqs createsubdb afsa_database_clu afsa_database afsa_database_clu_rep\n",
    "mmseqs convert2fasta afsa_database_clu_rep afsa_database_clu_rep.fasta   \n",
    "\n",
    "echo 'mod0-0.5' | awk -F '\\t' '{print $1}' *.tsv | sort | uniq -c | wc -l | cat\n",
    "\n",
    "conda deactivate\n",
    "\n",
    "############################\n",
    "# Muscle and CLustalo Alignment\n",
    "############################\n",
    "mkdir $OUT/AFSA_family_cluster/muscle\n",
    "\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate phylophlan\n",
    "\n",
    "# Add ancient AFSA genes and the cluster representatives in a file\n",
    "cat $ANCIENT_AFSA/*.faa $ROOT_AFSA/*.faa $OUT/AFSA_family_cluster/mmseqs/afsa_database_clu_rep.fasta > $OUT/AFSA_family_cluster/muscle/cluster.fasta\n",
    "\n",
    "muscle -in $OUT/AFSA_family_cluster/muscle/cluster.fasta \\\n",
    "-phyiout $OUT/AFSA_family_cluster/muscle/alignment.phy \\\n",
    "-clwstrictout $OUT/AFSA_family_cluster/muscle/alignment.clw \\\n",
    "-out $OUT/AFSA_family_cluster/muscle/alignment.afa -maxiters 5\n",
    "\n",
    "muscle -in $OUT/AFSA_family_cluster/muscle/cluster.fasta \\\n",
    "-out $OUT/AFSA_family_cluster/muscle/alignment.afa -maxiters 5\n",
    "\n",
    "clustalo -i $OUT/AFSA_family_cluster/muscle/alignment.clw \\\n",
    "--percent-id --distmat-out=$OUT/AFSA_family_cluster/muscle/pim.txt --full --force\n",
    "\n",
    "conda deactivate \n",
    "\n",
    "############################\n",
    "# RAXML tree construction\n",
    "############################\n",
    "mkdir $OUT/AFSA_family_cluster/tree\n",
    "\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate phylophlan\n",
    "\n",
    "cp $OUT/AFSA_family_cluster/muscle/alignment.phy $OUT/AFSA_family_cluster/tree/\n",
    "cd $OUT/AFSA_family_cluster/tree\n",
    "\n",
    "raxmlHPC-PTHREADS-AVX2 -f a -x 12345 -p 12345 -# 50 -m PROTGAMMAAUTO -T 29 -s alignment.phy -n alignment.tree \n",
    "\n",
    "conda deactivate\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb46ec62",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step5**: Alignment of the core genes of the Butyrolactone BGC\n",
    "\n",
    "##########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a020c6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############################\n",
    "#Hashes and Directories\n",
    "############################\n",
    "\n",
    "# NOTE: Change directories in bash script accordingly \n",
    "\n",
    "# Directories: \n",
    "OUT=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Output\n",
    "BINS=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/BINS\n",
    "SCRIPT=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Scripts\n",
    "\n",
    "DREPGENOME=/Net/Groups/ccdata/databases/ncbi-ref-genomes/Chlorobiales/drep/dereplicated_genomes\n",
    "AFSAFAMILY=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/Interpro_A0A101JTJ5\n",
    "ROOT_AFSA=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/AFSA_outgroup/*.faa\n",
    "ANCIENT_AFSA_comb=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/AFSA_aa.faa\n",
    "EMN_AFSA=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/AFSA_aa/EMN001_AFSA.faa\n",
    "ANCIENT_CONTIGS=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/ancient_contigs_names_butyrolactone\n",
    "BGC_GENES=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Input/BGC_genes\n",
    "\n",
    "############################\n",
    "# Grab percent identities of AFSA/OXI/SKI genes in Butyrolactone and terpene core protein\n",
    "############################\n",
    "\n",
    "mkdir $OUT/BGC_GENES\n",
    "cd $OUT/BGC_GENES\n",
    "\n",
    "# Add ancient AFSA genes and the cluster representatives in a file\n",
    "for F in $BGC_GENES/* ; do\n",
    "N=$(basename $F ) ;\n",
    "cat $F/*.fasta.fna > all_$N.nt.fasta ;\n",
    "cat $F/*.fasta.faa > all_$N.aa.fasta ; \n",
    "done\n",
    "\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate /home/aibrahim/anaconda3/envs/phylogenetics\n",
    "\n",
    "for F in $OUT/BGC_GENES/*.fasta ; do\n",
    "N=$(basename $F .fasta ) ;\n",
    "muscle -in $F \\\n",
    "-phyiout $OUT/BGC_GENES/$N.alignment.phy \\\n",
    "-clwstrictout $OUT/BGC_GENES/$N.alignment.clw \\\n",
    "-out $OUT/BGC_GENES/$N.alignment.afa -maxiters 10 ;\n",
    "muscle -in $F \\\n",
    "-out $OUT/BGC_GENES/$N.alignment.afa -maxiters 10 ;\n",
    "clustalo -i $OUT/BGC_GENES/$N.alignment.clw --percent-id --distmat-out=$OUT/BGC_GENES/$N.pim.txt --full --force ;\n",
    "sed -i '1d' $OUT/BGC_GENES/$N.pim.txt;\n",
    "done\n",
    "\n",
    "conda deactivate "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1b14a949",
   "metadata": {},
   "source": [
    "##########\n",
    "\n",
    "**Step6**: Butyrolactone cluster synteny - CLINKER\n",
    "\n",
    "##########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57e7507f",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "############################\n",
    "#Hashes and Directories\n",
    "############################\n",
    "\n",
    "# NOTE: Change directories in bash script accordingly \n",
    "\n",
    "# Directories: \n",
    "OUT=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Output\n",
    "\n",
    "BUT_CONTIGS=/Net/Groups/ccdata/users/AIbrahim/ancientDNA/Deep-Evo/BGC/final-butyrolactone/Output/CLINKER/butyrolactone_contigs\n",
    "\n",
    "mkdir $OUT\n",
    "\n",
    "############################\n",
    "# Grab the Butyrolactone BGC containing `.gbk` '$BUT_CONTIGS' files and run CLINKER\n",
    "############################\n",
    "mkdir $OUT/CLINKER/butyrolactone\n",
    "cd $OUT/CLINKER/butyrolactone\n",
    "\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate clinker \n",
    "\n",
    "# CLINKER\n",
    "clinker $BUT_CONTIGS/*.gbk -p butyrolactone.html -j 30 -o butyrolactone.aln\n",
    "\n",
    "conda deactivate\n",
    "\n",
    "# Manually: adjust the synteny image manually using the .html"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
