## Overview of the folder `02-scripts`

This folder contains all the scripts and Jupyter notebooks to prepare and analyse the data.

1. `PREP`: preparing of the data

  - `PREP_download_from_ENA.Snakefile`: download the sequencing data from ENA with the help of the
    Python program [ffq](https://github.com/pachterlab/ffq)
  - `PREP_preprocessing_dentalcalculus_sequencing_data.Snakefile`: prepare input file for
    pre-processing the dental calculus sequecing data with the Nextflow pipeline
    [nf-core/eager](https://nf-co.re/eager) and extract and merge all non-human reads per individual
  - `PREP_preprocessing_publishedNeanderthalcalculus.Snakefile`: download the sequencing data of the
    four Neanderthal dental calculus samples published by Weyrich *et al.* (2017) from ENA,
    pre-process the sequencing data using the Nextflow pipeline
    [nf-core/eager](https://nf-co.re/eager) and extract the non-human reads per individual
  - `PREP_preprocessing_ElMiron_sediments.Snakefile`: download the sequencing data of the six El
    Miron sediment samples generated for this study from ENA, pre-process the sequencing data using
    the Nextflow pipeline [nf-core/eager](https://nf-co.re/eager) and extract the non-human reads
    per individual
  - `PREP_preprocessing_ElMiron_toebone.Snakefile`: download the shotgun sequencing data of the El
    Miron toe bone sample generated by Fu et al. (2016), but unpublished, from ENA, pre-process the
    sequencing data using the Nextflow pipeline [nf-core/eager](https://nf-co.re/eager) and extract
    the non-human reads per individual
  - `PREP_preprocessing_lab_negcontrols.Snakefile`: download the sequencing data of the extraction
    and library negative controls that were run alongside the dental calculus and sediment samples
    of this study from ENA, pre-process the sequencing data using the Nextflow pipeline
    [nf-core/eager](https://nf-co.re/eager) and extract the non-human reads per individual

2. `ASMB`: analyses related to the *de novo* assembly and the non-reference binning

  - `ASMB_denovo_assembly_binning.Snakefile`: prepare the sample table and config file for
    assembling the ancient and modern DNA samples separately, run the assembly pipeline
    https://github.com/alexhbnr/ancient_metagenome_assembly, and summarise the results at contig and
    bin level
  - `ASMB_automaticRefinement.Snakefile`: prepare the sample table and config file for automatically
    refining the MAGs obtained from metaWRAP for the anceint and modern DNA samples separately, run
    the refinement pipeline https://github.com/alexhbnr/automatic_MAG_refinement, and summarise the
    results
  - `ASMB_rRNA_tRNA_search.Snakefile`: search for the presence of tRNAs and rRNAs in the MAGs that
    fulfill the completeness and contamination estimates for a high-quality MAG following MIMAG
  - `ASMB_dereplication.Snakefile`: use dRep to dereplicate the MAGs into representative genomes and
    compare these representative genomes against the HOMD for close matches
  - `ASMB_taxonomic_profiling_MAGs.Snakefile`: generate the taxonomic profiles of the representative
    MAGs using the GTDBTK and PhyloPhlAn3's metagenomic mode

3. `QUAL`: analyses related to the data quality

  - `QUAL_fragmentlengths.Snakefile`: summarise the DNA molecule fragment length across the
    sequencing data generated on 2x 75 bp paired-end Illumina sequencing runs for all samples
  - `QUAL_freeBayes_nomismatches_MEGAHIT.Snakefile`: evaluate the number of substitutions (per
    type), the minor allele frequencies of the variant calls, and the number of affected contigs for
    which freeBayes inferred a different major allele than it was exported by MEGAHIT from the
    assembly graph
  - `QUAL_pyDamage_summary.Snakefile`: summarise the pyDamage results regarding the fraction of
    contigs that are considered "ancient" by pyDamage and the distribution of the predicted accuracy
  - `QUAL_refalignment_Chlorobiaceae.Snakefile`: align the short-read sequencing data of all dental
    calculus samples against the contigs of EMN001 and evaluate the number of aligned reads against
    the contigs that were assigned to the Chlorobiaceae MAG
  - `QUAL_aDNA_damage_MAGs.Snakefile`: determine the amount of aDNA by quantifying the frequency of
    the substitutions for all five samples with HQ Chlorobiaceae MAGs against the genomes of the
    species *Chlorobium limicola* and Abot439

4. `REFG`: analyses related to constructing genomes using reference genomes

  - `REFG_Chlorobiaceae_genomes.Snakefile`: reconstruct the Chlorobiaceae genome for all samples
    without strong evidence of Chlorobiaceae DNA using the EMN001 Chlorobiaceae MAG as the reference

5. `PHYL`: analyses related to phylogenetics

  - `PHYL_Flexilinea_proteinalignment.Snakefile`: reconstruct the phylogeny of the family
    Anaerolineaceae using the ancient MAGs assigned to the genus Flexilinea based on the protein
    sequences of the UniProt genes associated with the species *Flexilinea flocculi*

6. `PUBL`: preparing the final figures and tables

  - `PUBL_Figure1.Snakefile`: plot an overview map of all samples, the DNA molecule length
    and contig length distributions for the samples EMN001, PES001, and PLV001, the overview of the
    reconstructed MAGs for the five samples with a HQ Chlorobiaceae MAG, and the aDNA damage for two
    MAGs of EMN001
  - `PUBL_Dataset_S1.Snakefile`: combine the tables containing the information about the samples,
    the sequencing data, and the number of DNA molecules available for *de novo* assembly into a
    single XSLX file
  - `PUBL_Dataset_S2.Snakefile`: combine the tables containing the results about the de novo
    assembly of the ancient and modern dental calculus samples
  - `PUBL_Dataset_S3.Snakefile`: combine the tables containing the number of MAGs after binning and
    the quality evaluation before and after the automatic refinement of the metaWRAP refined MAGs
  - `PUBL_Dataset_S4.Snakefile`: combine the tables containing the results on the alignment of the
    short-read sequencing data against the contigs assembled from EMN001
  - `PUBL_FigureS_freeBayescorrection.Snakefile`: plot the supplementary figure summarising the
    results of the substitution types and minor allele frequencies that were observed when
    correcting the contig sequences using freeBayes
  - `PUBL_FigureS_denovoassembly_metaQUAST_calN50.Snakefile`: plot the supplementary figure
    summarising the evaluation of the *de novo* assembly performance of the ancient and modern
    dental calculus samples
  - `PUBL_FigureS_pyDamage_contigs.Snakefile`: plot the supplementar figure summarising the results
    of pyDamage on contig level
  - `PUBL_FigureS_MAGquality_prefilter.Snakefile`: plot the quality analysis of the MAGs refined by
    metaWRAP estimated using checkM, BUSCO, and GUNC
  - `PUBL_FigureS_MAGquality_postfilter.Snakefile`: plot the quality analysis of the MAGs refined by
    metaWRAP estimated using checkM and GUNC
  - `PUBL_FigureS_MAG_taxonomicclassification.Snakefile`: plot the summary of the taxonomic
    assignment of the MAGs against the GTDB
  - `PUBL_FigureS_aDNA_damage_Chlorobium_Flexilinea.Snakefile`: plot the summary of the ancient DNA
    damage of the samples with a HQ Chlorobium MAG against the representative MAG of the genus
    Chlorobium and Flexilinea of EMN001

X. `ENVS`: conda environments for the use with Snakemake

  - `ENVS_bioawk.yaml`: [bioawk](https://github.com/lh3/bioawk)
  - `ENVS_bowtie2.yaml`: [bowtie2](https://github.com/BenLangmead/bowtie2)
  - `ENVS_damageprofiler.yaml`: [damageprofiler](https://github.com/Integrative-Transcriptomics/DamageProfiler)
  - `ENVS_dRep.yaml`: [dRep](https://github.com/MrOlm/drep)
  - `ENVS_entrez.yaml`: [entrez](https://www.ncbi.nlm.nih.gov/books/NBK179288/)
  - `ENVS_infernal.yaml`: [INFERNAL](http://eddylab.org/infernal/)
  - `ENVS_fastp.yaml`: [fastp](https://github.com/OpenGene/fastp)
  - `ENVS_PhyloPhlAn3.yaml`: [PhyloPhlAn3](https://github.com/biobakery/phylophlan)
  - `ENVS_picard.yaml`: [picard](https://broadinstitute.github.io/picard)
  - `ENVS_samtools.yaml`: [samtools](https://github.com/samtools/samtools)
  - `ENVS_trnascanse.yaml`: [tRNAscan-SE](https://github.com/UCSC-LoweLab/tRNAscan-SE)
